pkg/core/mempool/mempool.go

```diff
@@ -447,16 +448,42 @@ func (m Mempool) processGetMempoolTxsBySizeRequest(r rpcbus.Request) (interface{
 	}
 
 	txs := make([]transactions.ContractCall, 0)
+	gasLimit := config.Get().State.BlockGasLimit
+	// The slippageGasLimit is the threshold that consider the "estimated gas
+	// spent" acceptable even if it exceeds the strict GasLimit. This is
+	// required to avoid to iterate the whole meempol until it fit perfectly
+	// the block GasLimit
+	slippageGasLimit := gasLimit + gasLimit/10
 
 	var totalSize uint32
+	var totalGas uint64
 
 	err := m.verified.RangeSort(func(k txHash, t TxDesc) (bool, error) {
 		var done bool
+
+		decoded, err := t.tx.Decode()
+		if err != nil {
+			// Cannot decode, skip the tx.
+			// This should never happen, keeping `err` to log it properly`
+			return false, err
+		}
+
+		totalGas += decoded.EstimatedGasSpent()
+		if totalGas > slippageGasLimit {
+			// Total gas exceeded the slippage threshold, skip the tx
+			return false, nil
+		}
+
+		// If the totalGas exceeded the gasLimit (but still below the
+		// slippage threshold) stop to iterate the mempool
+		done = totalGas >= gasLimit
+
 		totalSize += uint32(t.size)
 
 		if totalSize <= maxTxsSize {
 			txs = append(txs, t.tx)
 		} else {
+			// it the totalSize exceeds the limit, stop to iterate the mempool
```

Comment:
Maybe this last bit could just be

```go
		if totalSize <= maxTxsSize {
			txs = append(txs, t.tx)
		}

		// We're done when the totalGas exceeds the gasLimit or
		// totalSize exceeds the tx size limit
		done = totalGas >= gasLimit || totalSize > maxTxsSize
```

Reply:
not exactly 

If we exceed the size, we must not include the tx
If we exceed the gas limit (but not the slippage) we have to include it

This is why I've split the checks

Reply:
Refactored right now to improve readability